src/renderer/src/pass/forward.rs

```diff
@@ -82,16 +83,36 @@ pub static FRAGMENT_SRC: &'static [u8] = b"
     out vec4 o_Color;
 
     void main() {
-        vec4 kd = texture(t_Kd, v_In.TexCoord);
         vec4 color = texture(t_Ka, v_In.TexCoord);
+        vec4 lighting = vec4(0.0);
+        vec4 normal = vec4(normalize(v_In.Normal), 0.0);
+
         for (int i = 0; i < u_LightCount; i++) {
-            vec4 delta = light[i].center - v_In.Position;
-            float dist = length(delta);
-            float inv_dist = 1. / dist;
-            vec4 light_to_point_normal = delta * inv_dist;
-            float intensity = dot(light[i].propagation.xyz, vec3(1., inv_dist, inv_dist * inv_dist));
-            color += kd * light[i].color * intensity * max(0, dot(light_to_point_normal, vec4(v_In.Normal, 0.)));
+            // Calculate diffuse light
+            vec4 lightDir = normalize(light[i].center - v_In.Position);
+            float diff = max(dot(lightDir, normal), 0.0);
+            vec4 diffuse = diff * light[i].color;
+
+            // Calculate specular light
+            vec4 specular = vec4(0.0);
+            if (diff > 0.0) {
```

Comment:
I don't think this is correct. While physically obvious, it doesn't quite work like that for blinn-phong model, which can have a slight specular effect on areas unaffected by the diffuse (because it's an approximation, after all). So I suggest just computing the diffuse and specular terms independently and combining them. Otherwise, you may see a sharp cut-off of the specular for some cases.